crates/proto/src/dnssec/dnssec_dns_handle/mod.rs

```diff
@@ -1337,77 +1342,142 @@ impl RrsigValidity {
+        return proof_log_yield(
+            Proof::Bogus,
+            query.name(),
+            "nsec1",
+            "unsupported response code",
+        );
     }
 
-    let verify_nsec_coverage = |query_name: &Name| -> bool {
-        nsecs.iter().any(|(nsec_name, nsec_data)| {
-            // the query name must be greater than nsec's label (or equal in the case of wildcard)
-            query_name >= nsec_name && {
-                // the query name is less than the next name
-                // or this record wraps the end, i.e. is the last record
-                query_name < nsec_data.next_domain_name()
-                    || nsec_data.next_domain_name() < nsec_name
-            }
-        })
+    let handle_matching_nsec = |type_set: &RecordTypeSet,
+                                message_secure: &str,
+                                message_record_exists: &str,
+                                message_name_exists| {
+        if type_set.contains(query.query_type()) || type_set.contains(RecordType::CNAME) {
+            proof_log_yield(Proof::Bogus, query.name(), "nsec1", message_record_exists)
+        } else if response_code == ResponseCode::NoError {
+            proof_log_yield(Proof::Secure, query.name(), "nsec1", message_secure)
+        } else {
+            proof_log_yield(Proof::Bogus, query.name(), "nsec1", message_name_exists)
+        }
     };
 
-    // continue to validate there is no wildcard
-    if !verify_nsec_coverage(query.name()) {
-        return proof_log_yield(Proof::Bogus, query.name(), "nsec1", "no wildcard");
+    // Look for an NSEC record that matches the query name first. If such a record exists, then the
+    // query type and CNAME must mot be present at this name.
+    if let Some((_, nsec_data)) = nsecs.iter().find(|(name, _)| query.name() == *name) {
+        return handle_matching_nsec(
+            nsec_data.type_set(),
+            "direct match",
+            "direct match, record should be present",
+            "nxdomain when direct match exists",
+        );
     }
 
-    // validate ANY or *.domain record existence
+    if !soa_name.zone_of(query.name()) {
+        return proof_log_yield(
+            Proof::Bogus,
+            query.name(),
+            "nsec1",
+            "SOA record is for the wrong zone",
+        );
+    }
 
-    // we need the wildcard proof, but make sure that it's still part of the zone.
-    let wildcard = query.name().base_name();
-    let wildcard = if soa_name.zone_of(&wildcard) {
-        wildcard
-    } else {
-        soa_name.clone()
+    let Some((covering_nsec_name, covering_nsec_data)) =
+        nsecs.iter().find(|(nsec_name, nsec_data)| {
+            soa_name.zone_of(nsec_name)
+                && query.name() > nsec_name
+                && (query.name() < nsec_data.next_domain_name()
+                    || nsec_data.next_domain_name() == soa_name)
+        })
```

Comment:
This construct is used here and below in the empty-non-terminal section -- suggest extracting it to make the intent clearer and reduce duplication.

```suggestion
        nsecs.iter().find(|(nsec_name, nsec_data)|
            nsec_covering_record(
                &soa_name,
                &query.name(),
                &nsec_name,
                &nsec_data.next_domain_name(),
            )
        )
```

Reply:
Done, thanks!